File: backend/app/routers/watchlist.py
from fastapi import APIRouter, Depends, status
from sqlalchemy.ext.asyncio import AsyncSession
from app.core.database import get_db
from app.core.deps import get_current_user
from app.models.user import User
from app.schemas.watchlist import AddWatchlistItem, WatchlistItemResponse
from app.services.watchlist_service import (
    get_watchlist,
    add_to_watchlist,
    remove_from_watchlist,
)
from uuid import UUID


router = APIRouter(prefix="/watchlist", tags=["watchlist"])


@router.get("", response_model=list[WatchlistItemResponse])
async def list_watchlist(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    items = await get_watchlist(db, current_user.id)
    return [WatchlistItemResponse.model_validate(item) for item in items]


@router.post("", response_model=WatchlistItemResponse)
async def add_item(
    data: AddWatchlistItem,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    item = await add_to_watchlist(db, current_user.id, data)
    return WatchlistItemResponse.model_validate(item)


@router.delete("/{item_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_item(
    item_id: UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    await remove_from_watchlist(db, current_user.id, item_id)

File: frontend/src/api/watchlist.ts
import api from './client';

export interface WatchlistItem {
  id: string;
  symbol: string;
  notes: string;
  created_at: string;
}

export const getWatchlist = async () => {
  const res = await api.get<WatchlistItem[]>('/watchlist');
  return res.data;
};

export const addToWatchlist = async (symbol: string, notes: string = '') => {
  const res = await api.post<WatchlistItem>('/watchlist', { symbol, notes });
  return res.data;
};

export const removeFromWatchlist = async (id: string) => {
  await api.delete(`/watchlist/${id}`);
};